src/django_components/expression.py

```diff
@@ -1,50 +1,55 @@
-from typing import Any, Dict, List, Mapping, Optional, Union
+from typing import Any, Dict, List, Mapping, Optional, Tuple, Union
 
-from django.template import Context
+from django.template import Context, TemplateSyntaxError
 from django.template.base import FilterExpression, Parser
 
+Expression = Union[FilterExpression]
+RuntimeKwargsInput = Dict[str, Expression]
+RuntimeKwargPairsInput = List[Tuple[str, Expression]]
 
-class AggregateFilterExpression:
-    def __init__(self, dict: Dict[str, FilterExpression]) -> None:
-        self.dict = dict
 
+class RuntimeKwargs:
+    def __init__(self, kwargs: RuntimeKwargsInput) -> None:
+        self.kwargs = kwargs
 
-Expression = Union[FilterExpression, AggregateFilterExpression]
+    def resolve(self, context: Context) -> Dict[str, Any]:
+        resolved_kwargs = safe_resolve_dict(context, self.kwargs)
+        return process_aggregate_kwargs(resolved_kwargs)
 
 
-def resolve_expression_as_identifier(
-    context: Context,
-    fexp: FilterExpression,
-) -> str:
-    resolved = fexp.resolve(context)
-    if not isinstance(resolved, str):
-        raise ValueError(
-            f"FilterExpression '{fexp}' was expected to resolve to string, instead got '{type(resolved)}'"
-        )
-    if not resolved.isidentifier():
-        raise ValueError(
-            f"FilterExpression '{fexp}' was expected to resolve to valid identifier, instead got '{resolved}'"
-        )
-    return resolved
+class RuntimeKwargPairs:
+    def __init__(self, kwarg_pairs: RuntimeKwargPairsInput) -> None:
+        self.kwarg_pairs = kwarg_pairs
 
+    def resolve(self, context: Context) -> List[Tuple[str, Any]]:
+        resolved_kwarg_pairs: List[Tuple[str, Any]] = []
+        for key, kwarg in self.kwarg_pairs:
+            resolved_kwarg_pairs.append((key, kwarg.resolve(context)))
 
-def safe_resolve_list(args: List[Expression], context: Context) -> List:
-    return [safe_resolve(arg, context) for arg in args]
+        return resolved_kwarg_pairs
 
 
-def safe_resolve_dict(
-    kwargs: Union[Mapping[str, Expression], Dict[str, Expression]],
-    context: Context,
-) -> Dict:
-    return {key: safe_resolve(kwarg, context) for key, kwarg in kwargs.items()}
+def is_identifier(value: Any) -> bool:
+    if not isinstance(value, str):
+        return False
+    if not value.isidentifier():
+        return False
+    return True
+
+
+def safe_resolve_list(context: Context, args: List[Expression]) -> List:
+    return [arg.resolve(context) for arg in args]
 
 
-def safe_resolve(context_item: Expression, context: Context) -> Any:
```

Comment:
Turns out `safe_resolve()` wasn't really needed, and instead we can call `.resolve()` directly on the objects that support it.